ssz/src/main/java/tech/pegasys/artemis/util/backing/view/AbstractCompositeViewRead.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.view;
+
+import tech.pegasys.artemis.util.backing.CompositeViewRead;
+import tech.pegasys.artemis.util.backing.ViewRead;
+import tech.pegasys.artemis.util.backing.cache.ArrayCache;
+import tech.pegasys.artemis.util.backing.cache.IntCache;
+import tech.pegasys.artemis.util.backing.tree.TreeNode;
+import tech.pegasys.artemis.util.backing.type.CompositeViewType;
+
+/**
+ * Base backing view class for immutable composite views (lists, vectors, containers)
+ *
+ * <p>It caches it's child view instances so that if the underlying tree nodes are not changed (in
+ * the corresponding mutable classes) the instances are not recreated from tree nodes on later
+ * access.
+ *
+ * <p>Thought internally this class has a mutable cache it may be thought of as immutable instance
+ * and used safely across threads
+ *
+ * @param <ChildType> the type of children. For heterogeneous composites (like container) this type
+ *     would be just generic {@link ViewRead}
+ */
+public abstract class AbstractCompositeViewRead<ChildType extends ViewRead>
+    implements CompositeViewRead<ChildType> {
+
+  private IntCache<ChildType> childrenViewCache;
+  private final int sizeCache;
+  private final CompositeViewType type;
+  private final TreeNode backingNode;
+
+  /** Creates an instance from a type and a backing node */
+  protected AbstractCompositeViewRead(CompositeViewType type, TreeNode backingNode) {
+    this(type, backingNode, null);
+  }
+
+  /**
+   * Creates an instance from a type and a backing node Optionally the view instances cache can be
+   * supplied for optimization to shortcut children views creation from backing nodes. The cache
+   * should correspond to the supplied backing tree.
+   */
+  protected AbstractCompositeViewRead(
+      CompositeViewType type, TreeNode backingNode, IntCache<ChildType> cache) {
```

Comment:
(nit) We tend to use `Optional`'s explicitly rather than allow nullable fields / parameters.
```suggestion
      CompositeViewType type, TreeNode backingNode, Optional<IntCache<ChildType>> cache) {
```

Reply:
Hm, seems like the `Optional` was generally intended to be used just for return types and not for params/fields but personally for me that doesn't sound like a dogma and denoting optional argument this way may be a good practice
Anyway this class has also a constructor `AbstractCompositeViewRead(CompositeViewType type, TreeNode backingNode)` so I would better just make the `cache` param non-null